Chat Refactor 1: Compoundable Adapters

designatednerd opened the pull request:
Looking at incoming designs, there's a strong we're going to need to show differing viewTypes at differing places. We had some architecture around this in `RecentAdapter`, but it involved a lot of direct handling of what went where when which seemed *really* complicated. 

 I thought moving to a compoundable adapter architecture so we can just pull sections in and out (or easily hide them) at will, would be helpful. Each section would be a homogenous list of items, but together they can be compounded into a single adapter which ultimately feeds the RecyclerView. 

This allows logic to be broken out per-adapter so that each adapter can focus on formatting item views of its own type, and then adapters to be quickly snapped together in order to form these heterogeneous RecyclerViews. 

There's a big piece of this where I'm super-fighting Kotlin's type system, and would appreciate any advice on how **not** to do that. Basically, type erasure with Kotlin (and IIRC Java) generics means that if I want to have a `<VH: RecyclerView.ViewHolder>` type, I can't use `as?` casting with `VH` to validate that the type is correct later - only with the actual subclass being used in the concrete implementation. Swift's generics don't type-erase, so this isn't a problem I've dealt with in prior implementations of this same concept, except for iOS. 

I also threw in a whole bunch of tests to try and make it clearer what I'm trying to do (and backstop me as I work through this).

Any thoughts on this would be most appreciated.

martintoften commented on app/src/main/java/com/toshi/view/adapter/CompoundAdapter.kt:
```diff
@@ -0,0 +1,94 @@
+package com.toshi.view.adapter
```
License is missing here

csdodd commented on app/src/main/java/com/toshi/view/adapter/BaseCompoundableAdapter.kt:
```diff
@@ -0,0 +1,144 @@
+        itemsToRemove.add(removedItem)
+    }
+
+    // UNDO! UNDO!
```
Can we keep comments to a minimum and only when they add value? This comment can be removed.

csdodd replied:
Whilst we're on the subject. I am not sure how much value splitting classes into various sections by these comments has at all.

I foresee issues whereby code is incorrectly placed relative to a comment causing confusion for future devs. Or that some function doesn't neatly pass into one "section", or straddles multiple "sections". Which begs the question, what value are these comments adding at all?

If you need these sections to help understand the class it is a fairly clear sign that the class may be too long and may require refactoring. Since this is not a pattern followed elsewhere I would strongly recommend you remove these "section" comments unless you have a strong argument for keeping them.

designatednerd replied:
Fair enough - this is a thing we do in iOS because it makes it easier to jump between sections because of how the IDE allows you to look at different sets of methods, but I'm fine with removing it here. I'm just kind of used to using it so I think I did it automatically

csdodd commented on app/src/main/java/com/toshi/view/adapter/CompoundAdapter.kt:
```diff
@@ -0,0 +1,245 @@
+/**
+An interface which allows us to avoid generic hell but still combine different types
+ */
+interface CompoundableAdapter {
```
This comment should follow the Javadoc format, then the IDE will be able to render it during automcompletion.

martintoften commented on app/src/main/java/com/toshi/view/adapter/CompoundAdapter.kt:
```diff
@@ -0,0 +1,94 @@
+    // Add observers when data set changes
+    private fun currentSectionAdapter(position: Int): RecyclerView.Adapter<RecyclerView.ViewHolder> {
+        var previousCount = 0
+        adapters.forEach { adapter ->
```
You can remove `adapter` and use `it` instead. It's pretty easy to know what `it` is in this context

martintoften commented on app/src/main/java/com/toshi/view/adapter/CompoundAdapter.kt:
```diff
@@ -0,0 +1,94 @@
+        return sectionAdapter.onCreateViewHolder(parent, 0)
+    }
+
+
```
Is there a reason why you have an extra line here?

martintoften commented on app/src/main/java/com/toshi/view/adapter/CompoundAdapter.kt:
```diff
@@ -0,0 +1,94 @@
+        }
+    }
+
+
```
Same here

martintoften commented on app/src/main/java/com/toshi/view/adapter/ConversationAdapter.kt:
```diff
@@ -0,0 +1,104 @@
+package com.toshi.view.adapter
```
Missing license

martintoften commented on app/src/main/java/com/toshi/view/adapter/ConversationAdapter.kt:
```diff
@@ -0,0 +1,104 @@
+        private val onItemLongClickListener: OnItemClickListener<Conversation>
+    ): RecyclerView.Adapter<ThreadViewHolder>() {
+
+    override fun getItemCount(): Int {
```
This could be a one-liner

martintoften commented on app/src/main/java/com/toshi/view/adapter/ConversationAdapter.kt:
```diff
@@ -0,0 +1,104 @@
+            }
+        } catch (ex: IOException) {
+            LogUtil.w("Error parsing SofaMessage. $ex")
+
```
This empty line shouldn't be here

martintoften commented on app/src/main/java/com/toshi/view/adapter/ConversationAdapter.kt:
```diff
@@ -16,18 +16,23 @@ import com.toshi.view.adapter.listeners.OnItemClickListener
+        this.conversations = conversations
+    }
+
+
```
I see that you add extra empty lines. Is there a reason for it?

martintoften commented on app/src/main/java/com/toshi/extensions/RecyclerViewAdapterExtensions.kt:
```diff
@@ -0,0 +1,8 @@
+package com.toshi.extensions
```
Missing license

csdodd commented on app/src/main/java/com/toshi/view/adapter/CompoundAdapter.kt:
```diff
@@ -0,0 +1,245 @@
+
+    // FIGURING OUT WHERE WE ARE
+
+    private fun totalItemsBeforeSection(sectionIndex: Int): Int {
```
We aim to have code that can be read from top to bottom.

So since this function is called directly by the functions below, it should be physically moved to be placed under those functions.

Admittedly it is not always obvious, easy, or possible to achieve this but as a general rule -- code should be calling functions that are physically lower in the class file.

designatednerd replied:
ah okay - I'm used to "if you're going to call it later, make sure people know what the fuck it's doing before you call it" structuring. will update 👍

martintoften commented on app/src/main/java/com/toshi/extensions/RecyclerViewAdapterExtensions.kt:
```diff
@@ -0,0 +1,8 @@
+
+
+fun <VH: RecyclerView.ViewHolder>RecyclerView.Adapter<VH>.isEmpty(): Boolean {
+    return (this.itemCount == 0)
```
You can drop the parentheses here

designatednerd replied:
i actually kind of prefer them in terms of readability - makes it clearer that you're returning the result of what's in the parens. 

If you guys prefer without the parens I'll remove them though

csdodd replied:
Remove them (https://github.com/bakkenbaeck/android-handbook/blob/master/java-style-guide.md#dont-wrap-return-statements-in-parentheses)

martintoften commented on app/src/main/java/com/toshi/view/adapter/ConversationAdapter.kt:
```diff
@@ -14,16 +14,45 @@ import com.toshi.util.logging.LogUtil
+        return onCreateViewHolder(parent, viewType)
+    }
+
+    override fun genericItemCount(): Int {
```
This could be a one-liner

martintoften commented on app/src/main/java/com/toshi/view/adapter/ConversationRequestAdapter.kt:
```diff
@@ -34,7 +34,22 @@ class ConversationRequestAdapter(
+        return onCreateViewHolder(parent, viewType)
+    }
+
+    override fun genericItemCount(): Int {
```
Could be a one-liner

martintoften commented on app/src/main/java/com/toshi/view/adapter/CompoundAdapter.kt:
```diff
@@ -22,19 +22,23 @@ class CompoundAdapter(
-    private fun adapterAtIndex(index: Int): CompoundableAdapter {
-        return adapters[index]
+    fun indexOf(adapter: CompoundableAdapter): Int? {
+        if (adapters.contains(adapter)) {
```
Can can write this as:

```
return if (adapters.contains(adapter)) {
    adapters.indexOf(adapter)
} else {
    null
}
```

Better to write `return` once

martintoften commented on app/src/main/java/com/toshi/view/adapter/BaseCompoundableAdapter.kt:
```diff
@@ -0,0 +1,76 @@
+        this.parent = parent
+    }
+
+    fun itemAt(index: Int): T {
```
This could be a one-liner

designatednerd replied:
I do prefer it but not strongly enough to argue about it. 😄 moving what I can to one-liners.

csdodd replied:
We could be more explicit in the styling guide, but the general rule is if it fits on one-line make it a one-liner (https://github.com/bakkenbaeck/android-handbook/blob/master/kotlin-style-guide.md#simplify-functions)

martintoften commented on app/src/main/java/com/toshi/view/fragment/toplevel/RecentFragment.kt:
```diff
@@ -132,7 +132,7 @@ class RecentFragment : TopLevelFragment() {
 
         recents.apply {
-            layoutManager = LinearLayoutManager(context)
+            layoutManager =  LinearLayoutManager(context)
```
An extra whitespace here

martintoften commented on app/src/main/java/com/toshi/util/keyboard/SOFAMessageFormatter.kt:
```diff
@@ -0,0 +1,70 @@
+package com.toshi.util.keyboard
```
Missing license

martintoften commented on app/src/main/java/com/toshi/util/keyboard/SOFAMessageFormatter.kt:
```diff
@@ -0,0 +1,70 @@
+ */
+class SOFAMessageFormatter() {
+
+    private var localUser = BaseApplication
```
You you use the `lateinit` modifier here and set `localUser` when possible. I think we should pass `User` in a constructor(as you did) or in a set function. `SOFAMessageFormatter` shouldn't know about `BaseApplication`. I know that we have done this in other places, but it's not optimal

designatednerd replied:
This one I actually disagree on - part of the reason I isolated this into a class and just made the dependency-injectable constructor for testing (which, to be fair, I haven't done yet since this PR is huge enough and also I don't know the formatting rules) is that we were doing different backflips in different places just to get the user for the formatting, when the callers didn't need the user for any other reason. Here, the user is being grabbed because it's actually needed. It's never actually needed in the adapters directly. 

I sort of get the argument that this class doesn't really need to know about `BaseApplication`, but the counter-argument is: Why should two adapters know it exists instead?

martintoften replied:
It's not a big deal. I think the adapters should know about the user because the SOFAMessageFormatter must have `User` to work properly and it shouldn't know how to get it.

martintoften commented on app/src/main/java/com/toshi/view/adapter/RecentAdapter.java:
```diff
@@ -114,7 +115,7 @@ public void onBindViewHolder(final RecyclerView.ViewHolder holder, final int pos
             final ThreadViewHolder viewHolder = (ThreadViewHolder) holder;
             final Conversation conversation = (Conversation) conversationItem;
-            final String formattedLatestMessage = formatLastMessage(conversation.getLatestMessage());
+            final String formattedLatestMessage = (new SOFAMessageFormatter()).formatMessage(conversation.getLatestMessage());
```
This should be an instance variable. This will make a new `SOFAMessageFormatter` instance every time `onBindViewHolder` is invoked

designatednerd replied:
yeah i just put this one in temporarily while I was trying to get things to compile, I eventually got rid of this adapter completely

martintoften commented on app/src/main/java/com/toshi/view/adapter/BaseCompoundableAdapter.kt:
```diff
@@ -59,18 +100,45 @@ abstract class BaseCompoundableAdapter<VH: RecyclerView.ViewHolder, T>: Recycler
+        val snackbar = generateSnackbar(parentView)
+        snackbar.setAction(
+                R.string.undo,
+                { _ -> handleUndo(removedIndex, removedItem, parentView) }
```
You can remove `_` here

designatednerd replied:
ah nice - this is similar to the notation for "I am explicitly not handling this parameter" in swift. Swift closures won't compile if you don't at least explicitly discard an incoming parameter.

designatednerd replied:
also fwiw i checked to see if I had done this other places and this is allllll over the place in the code. Might be worth removing in a separate pull if you all don't want to use this style

csdodd replied:
Should be removed. You can make the changes in this PR or a later one, up to you. https://github.com/bakkenbaeck/android-handbook/blob/master/kotlin-style-guide.md#drop-the-left-part-of-the-lambda-expression-when-possible

martintoften replied:
We have to do it in java, if you have only one parameter you can drop it

martintoften commented on app/src/main/java/com/toshi/view/adapter/ConversationRequestAdapter.kt:
```diff
@@ -29,16 +29,33 @@ class ConversationRequestAdapter(
+    fun setConversations(conversations: List<Conversation>) {
+        this.conversations.clear()
+        this.conversations.addAll(conversations)
+        this.notifyDataSetChanged()
```
`this` is unnecessary when calling `notifyDataSetChanged`

designatednerd commented on app/src/main/java/com/toshi/view/adapter/ConversationAdapter.kt:
```diff
@@ -46,4 +48,15 @@ class ConversationAdapter(
     }
+
+    override fun deleteItem(item: Conversation) {
+        super.deleteItem(item)
```
The super call is actually unnecessary since the base implementation does nothing (I considered making it abstract but then everything would have to implement it, and that seemed like overkill.

designatednerd replied:
Does that address your concern here?

martintoften replied:
Sorry, I didn't mean the super call. I was thinking of the next line, but I can't see it now

designatednerd commented:
OK I think because I rebased and force-pushed (sorry, didn't see there were comments before I did that 🤦‍♀️ ) a couple of the "missing license" comments aren't resolving. 

However otherwise I think I've addressed all the comments. Let me know if you guys have more thoughts @csdodd @martintoften

designatednerd commented:
@martintoften You mean as a single commit? I try to keep them granular usually since if you push them all at once you should be able to review them with the "changes since you last viewed" thing on Github. Or is there something else you mean that I'm missing?

designatednerd commented:
@csdodd OK by you too?

martintoften commented:
No, I mean, when you fix everything in a single commit you add a new fixup commit which is linked to the commit you added fixes for. Then it's easier for me to review the fixes. After I've reviewed the fixes you can squash the fixup commits. When everything is squash I'll review again. Me and @csdodd used to do this.

designatednerd commented:
may god have mercy on us all
